[HGCal] TICL v3 major upgrade

felicepantaleo opened the pull request:
#### PR description:
This PR updates the TICL reconstruction to make it more robust against pile-up. 
In particular the following actions have been taken:

- a track seeded electromagnetic iteration has been introduced;
- the order of the iterations has been changed to TrkEM, EM unseeded, Trackseeded hadronic, hadronic unseeded
- layerClusters with a minimum size of three are used for pattern recognition
- seeding regions are sorted by pT and processed in this order
- EM iterations are limited to the first 30 layers
- a cut on the first layer id for the shower start has been introduced
- a cut on the number of missing layers has been introduced
- a cut on the longitudinal compactness of a trackster (sigmaPCA-z)
- a simple PF interpretation has been introduced
- the MIP iteration has been dropped 
#### PR validation:
This is the electron reconstruction with PU200 before this PR:
![gsfEle-E_by_genEle-E_PU200](https://user-images.githubusercontent.com/6595715/96882161-52f32280-147f-11eb-8744-311e2feb596b.png)

This is after the PR:
![gsfEle-E_by_genEle-E_PU200 (3)](https://user-images.githubusercontent.com/6595715/96882313-7918c280-147f-11eb-8026-0bfeed2653b9.png)


### Single Particle noPU 
All samples with 6 energy steps = 10, 20, 50, 100, 200, 300 GeV, eta = 1.8 (HGCAL center)


![image](https://user-images.githubusercontent.com/6595715/98122024-568b9e00-1eb0-11eb-8339-fa3f48c160d1.png)

![image](https://user-images.githubusercontent.com/6595715/98122057-60150600-1eb0-11eb-914d-acd88a340d8a.png)

![image](https://user-images.githubusercontent.com/6595715/98122122-715e1280-1eb0-11eb-816e-2410a8eb5b0b.png)

![image](https://user-images.githubusercontent.com/6595715/98122157-7d49d480-1eb0-11eb-927d-5f39d9b90caa.png)

![image](https://user-images.githubusercontent.com/6595715/98122188-876bd300-1eb0-11eb-8885-36437d89c565.png)

#### More info
you can find the latest physics results with this pull request described in the talks by HGCAL and Jets/MET at the HLT Upgrade workshop:
https://indico.cern.ch/event/962025/#4-hgcal
https://indico.cern.ch/event/962025/#6-jetsmet

### Timing Report

Timing with PR:
https://fpantale.web.cern.ch/fpantale/circles/web/piechart.html?local=false&dataset=TICLv3_PR31906&resource=time_thread&colours=default&groups=reco_PhaseII&threshold=0

felicepantaleo committed:
Created a ClusterFilter by Algo, Size and Layer Range

felicepantaleo committed:
include an Electromagnetic TrackSeeded iteration as first iteration

felicepantaleo committed:
Add the range of layers to the filter parameters

felicepantaleo committed:
Inject tracksters from the track seeded electromagnetic interation in the merged collection

felicepantaleo committed:
create empty multiclusters

rovere committed:
Add TrkEM tracksters into the Merge collection

rovere committed:
Select trackster-PID also on EM/Total trackster energy ratio

rovere committed:
Fix merge conflict.

felicepantaleo committed:
Fix input collection for TRK iteration to be the output of EM iteration

felicepantaleo committed:
Sorting seeding regions by track momentum

felicepantaleo committed:
Use only clusters with minimum size of 3

felicepantaleo committed:
Filter only with a maxLayerId in TRKEM iteration.

felicepantaleo committed:
Remove MIP iteration

felicepantaleo committed:
Protect from empty multiclusters

felicepantaleo committed:
clang format

felicepantaleo committed:
Limit Global EM iteration to layer 30

felicepantaleo committed:
Limit track seeded iterations to tracks with at most 2 missing outer hits in the Tracker

felicepantaleo committed:
Add TrkEM products to EventContent

felicepantaleo committed:
Remove missing layers cut for EM based iterations

felicepantaleo committed:
Introducing a shower_start_max_layer cut

felicepantaleo committed:
raising shower_start_max_layer from 3 to 7 in EM iterations

felicepantaleo committed:
Count usage of layer clusters, only for those tracksters passing the selection

felicepantaleo committed:
Increasing missing outer hits from 3 to 5

felicepantaleo committed:
increase missing layer from 0 to 1 and loosen min_cos_theta to 0.961 in EM iterations

felicepantaleo committed:
tighten min_cos_theta from 0.961 to 0.978 in EM iterations

felicepantaleo committed:
Rename parameter missing_layers to skip_layers.

felicepantaleo committed:
Retuning of PR cuts in EM iterations

felicepantaleo committed:
Introducing two cuts: a max_missing_layers_in_trackster and a minimum number of different layers in a trackster

felicepantaleo committed:
Refactoring

rovere committed:
Update cuts

felicepantaleo committed:
Make max_missing_layers_in_trackster cut inclusive

felicepantaleo committed:
code format

felicepantaleo committed:
Move trackster energy calculation after the selections are applied

felicepantaleo committed:
Add a new max_longitudinal_sigmaPCA cut for EM iterations, with an initial value of 10

felicepantaleo committed:
Add a new cut: root_doublet_max_distance_from_seed for TrkEM iteration, initial value 0.05

felicepantaleo committed:
clang format

felicepantaleo committed:
fix trackster to seed association

felicepantaleo committed:
Setting  max_missing_layers_in_trackster to 2 from 0 in EM iterations

rovere committed:
Fix compilation error in C++

rovere committed:
Remove duplicate python parameter

felicepantaleo committed:
Sorting seeding regions by pT rather than p

felicepantaleo committed:
Added TICL PF interpretation.

felicepantaleo committed:
Feed PFTICL the correct InputTag

felicepantaleo committed:
Adapt event content

cmsbuild commented:
The code-checks are being triggered in jenkins.

felicepantaleo commented:
FYI
@rovere @fwyzard @trtomei @gouskos @hqucms @ebrondol @lecriste @gennai @Sam-Harper @SohamBhattacharya @missirol

felicepantaleo commented:
enable profiling

felicepantaleo commented:
@cmsbuild please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31906/19329

- This PR adds an extra 224KB to repository

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
A new Pull Request was created by  @felicepantaleo (Felice Pantaleo) for master.

It involves the following packages:

DataFormats/HGCalReco
RecoHGCal/Configuration
RecoHGCal/TICL
RecoParticleFlow/PFClusterProducer
Validation/HGCalValidation


@perrotta, @andrius-k, @kmaeshima, @ErnestaP, @kpedro88, @cmsbuild, @jfernan2, @fioriNTU, @slava77, @jpata can you please review it and eventually sign? Thanks.
@mmarionncern, @lecriste, @sethzenz, @bsunanda, @clelange, @riga, @cbernet, @vandreev11, @rovere, @lgray, @cseez, @apsallid, @sobhatta, @pfs, @deguio, @hatakeyamak, @seemasharmafnal this is something you requested to watch as well.
@silviodonato, @dpiparo, @qliphy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

hatakeyamak commented:
Thanks @felicepantaleo.
If you don't mind, for my info, can you expand a bit on "a simple PF interpretation has been introduced"? (or just point to some reference on this point)?

cmsbuild commented:
+1
Tested at: ff51d7e53eae43a38c5b38dd264a07bd4a8ceff1
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-814241/10216/summary.html
CMSSW: CMSSW_11_2_X_2020-10-21-2300
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

kpedro88 commented on RecoHGCal/TICL/plugins/TrackstersMergeProducer.cc:
```diff
@@ -168,185 +185,331 @@ void TrackstersMergeProducer::produce(edm::Event &evt, const edm::EventSetup &es
+            }
+            tmpCandidate.setPdgId(11 * track.charge());
+            float p = tracksterTotalRawPt * cosh(t.barycenter().eta());
+            tmpCandidate.setRawEnergy(p);
```
it seems like this line and the PDG ID are the only differences between this case and the `foundCompatibleTRK` case above. a lot of the duplication could be avoided. the same is likely also true for the other, similar block above.

felicepantaleo replied:
Since the PF interpretation is under active development, I would prefer that the factorization is postponed once this has become final and keep the code more readable (although longer).

kpedro88 commented on RecoHGCal/TICL/plugins/TrackstersMergeProducer.cc:
```diff
@@ -168,185 +185,331 @@ void TrackstersMergeProducer::produce(edm::Event &evt, const edm::EventSetup &es
+                                            gammaEnergy);
+            photonCandidate.setP4(gammaP4);
+            photonCandidate.addTrackster(edm::Ptr<ticl::Trackster>(trackstersMergedHandle, tmpIndex));
+            resultCandidates->push_back(photonCandidate);
```
copy could be avoided by using `emplace_back()` and then `back()`

kpedro88 commented on RecoHGCal/TICL/plugins/TrackstersMergeProducer.cc:
```diff
@@ -168,185 +185,331 @@ void TrackstersMergeProducer::produce(edm::Event &evt, const edm::EventSetup &es
+          for (auto otherTracksterIdx : trackstersTRKEMwithSameSeed) {
+            auto tmpIndex = (otherTracksterIdx != closestTrackster) ? otherTracksterIdx : mergedIdx;
+            TICLCandidate photonCandidate;
+            auto &otherTrackster = trackstersMergedHandle->at(tmpIndex);
```
`const auto&`

kpedro88 commented on RecoHGCal/TICL/plugins/TrackstersMergeProducer.cc:
```diff
@@ -168,185 +185,331 @@ void TrackstersMergeProducer::produce(edm::Event &evt, const edm::EventSetup &es
+
+  for (unsigned i = 0; i < trackstersTRK.size(); ++i) {
+    auto mergedIdx = indexInMergedCollTRK[i];
+    auto &t = trackstersTRK[i];  //trackster
```
`const auto&`

kpedro88 commented on RecoHGCal/TICL/plugins/TrackstersMergeProducer.cc:
```diff
@@ -168,185 +185,331 @@ void TrackstersMergeProducer::produce(edm::Event &evt, const edm::EventSetup &es
+        math::XYZTLorentzVector p4(
+            p * track.momentum().unit().x(), p * track.momentum().unit().y(), p * track.momentum().unit().z(), energy);
+        tmpCandidate.setP4(p4);
+        resultCandidates->push_back(tmpCandidate);
```
copy could be avoided by using `emplace_back()` and then `back()`

kpedro88 commented on RecoHGCal/TICL/plugins/TrackstersMergeProducer.cc:
```diff
@@ -168,185 +185,331 @@ void TrackstersMergeProducer::produce(edm::Event &evt, const edm::EventSetup &es
+      tmpCandidate.setRawEnergy(energy);
+      math::XYZTLorentzVector p4(track.momentum().x(), track.momentum().y(), track.momentum().z(), energy);
+      tmpCandidate.setP4(p4);
+      resultCandidates->push_back(tmpCandidate);
```
copy could be avoided by using `emplace_back()` and then `back()`

kpedro88 commented on RecoHGCal/TICL/plugins/TrackstersMergeProducer.cc:
```diff
@@ -168,185 +185,331 @@ void TrackstersMergeProducer::produce(edm::Event &evt, const edm::EventSetup &es
+  // for all general tracks (high purity, pt > 1), check if they have been used: if not, promote them as charged hadrons
+  for (unsigned i = 0; i < tracks.size(); ++i) {
+    auto const &track = tracks[i];
+    if (track.pt() > 1.f and track.quality(reco::TrackBase::highPurity) and track.missingOuterHits() < 5 and
```
a lot of magic numbers here: should be named constants, configurable, or maybe obtained from geometry

kpedro88 commented on RecoHGCal/TICL/plugins/TrackstersMergeProducer.cc:
```diff
@@ -168,185 +185,331 @@ void TrackstersMergeProducer::produce(edm::Event &evt, const edm::EventSetup &es
+      tmpCandidate.setRawEnergy(energy);
+      math::XYZTLorentzVector p4(track.momentum().x(), track.momentum().y(), track.momentum().z(), energy);
+      tmpCandidate.setP4(p4);
+      resultCandidates->push_back(tmpCandidate);
```
copy could be avoided by using `emplace_back()` and then `back()`

kpedro88 commented on Validation/HGCalValidation/src/HGVHistoProducerAlgo.cc:
```diff
@@ -1511,189 +1511,189 @@ void HGVHistoProducerAlgo::multiClusters_to_CaloParticles(const Histograms& hist
+            cPOnLayer[h.clusterId][lcLayerId].layerClusterIdToEnergyAndScore[mclId].second = FLT_MAX;
+            //cpsInMultiCluster[multicluster][CPids]
+            //Connects a multi cluster with all related caloparticles.
+            cpsInMultiCluster[mclId].emplace_back(std::make_pair<int, float>(h.clusterId, FLT_MAX));
```
`make_pair` is unnecessary with `emplace_back`

kpedro88 commented on Validation/HGCalValidation/src/HGVHistoProducerAlgo.cc:
```diff
@@ -1511,189 +1511,189 @@ void HGVHistoProducerAlgo::multiClusters_to_CaloParticles(const Histograms& hist
-        occurrencesCPinMCL[c]++;
+      //Loop through all rechits to count how many of them are noise and how many are matched.
+      //In case of matched rechit-simhit, he counts and saves the number of rechits related to the maximum energy CaloParticle.
+      for (auto& c : hitsToCaloParticleId) {
```
`auto` (get primitive types by value)

kpedro88 commented on Validation/HGCalValidation/src/HGVHistoProducerAlgo.cc:
```diff
@@ -1511,189 +1511,189 @@ void HGVHistoProducerAlgo::multiClusters_to_CaloParticles(const Histograms& hist
-        maxCPNumberOfHitsInMCL = c.second;
+      //Below from all maximum energy CaloParticles, he saves the one with the largest amount
+      //of related rechits.
+      for (auto& c : occurrencesCPinMCL) {
```
could use structured binding e.g. `auto [id, nhits] : occurrencesCPinMCL`

kpedro88 commented on Validation/HGCalValidation/src/HGVHistoProducerAlgo.cc:
```diff
@@ -1511,189 +1511,189 @@ void HGVHistoProducerAlgo::multiClusters_to_CaloParticles(const Histograms& hist
-      for (unsigned int j = 0; j < layers * 2; ++j) {
-        totalCPEnergyFromLayerCP = totalCPEnergyFromLayerCP + cPOnLayer[maxCPId_byEnergy][j].energy;
+      //Find the CaloParticle that has the maximum energy shared with the multicluster under study.
+      for (auto& c : CPEnergyInMCL) {
```
could use structured binding

kpedro88 reviewed (commented):
This review includes some suggestions to reduce CPU usage and code duplication, as well as a few other minor points. The performance profile from the PR test may provide more direction re: reducing CPU usage (if necessary).

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-814241/10216/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - ROOTFileChecks: Some differences in event products or their sizes found
 - Reco comparison results: 1071 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2544110
 - DQMHistoTests: Total failures: 2389
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2541699
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 34 files compared)
 - Checked 149 log files, 22 edm output root files, 35 DQM output files

kpedro88 commented:
enable profiling

kpedro88 commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.
Test Parameters:
  - **ENABLE_BOT_TESTS** = PROFILING

cmsbuild commented:
+1
Tested at: ff51d7e53eae43a38c5b38dd264a07bd4a8ceff1
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-814241/10222/summary.html
CMSSW: CMSSW_11_2_X_2020-10-22-1100
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-814241/10222/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - ROOTFileChecks: Some differences in event products or their sizes found
 - Reco comparison results: 1071 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2544110
 - DQMHistoTests: Total failures: 2389
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2541699
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 34 files compared)
 - Checked 149 log files, 22 edm output root files, 35 DQM output files

fwyzard commented:
Many of the comments are along the line of

> copy could be avoided by using emplace_back() and then back()


Is there any evidence that it actually gives better performance?
The proposed change would access the data through a pointer instead of on
the stack.

jfernan2 commented:
From DQM side, plots in HGCAL / TICLPFCandidates / 3  get now zero entries and HGCAL / HGCalValidator / ticlMultiClustersFromTrackstersMerge/Trk now get many Denominator null values
I guess this is just a matter of a new definition, but I wonder if it is expected. I have made the same question in the backport PR #31907

fwyzard commented:
About the suggestion to use

    const auto&

instead of

    auto&

I thought we had introduced a clang-tidy check to make this automatic.
Do I remember incorrectly ?

VinInn commented on RecoHGCal/TICL/plugins/PatternRecognitionbyCA.cc:
```diff
@@ -137,29 +155,105 @@ void PatternRecognitionbyCA<TILES>::makeTracksters(
+      lcIdAndLayer.emplace_back(i, layerId);
+    }
+    std::sort(uniqueLayerIds.begin(), uniqueLayerIds.end());
+    uniqueLayerIds.erase(std::unique(uniqueLayerIds.begin(), uniqueLayerIds.end()), uniqueLayerIds.end());
```
the fastest is to build a heap while pushing and then do the erase(unique());
std::set is by definition slower then a vector as it allocated node on the fly (no reserve)

felicepantaleo replied:
for a very small number of elements, the vector is the fastest option. We are talking of 20-30 elements here

felicepantaleo committed:
Address review comments

cmsbuild commented:
The code-checks are being triggered in jenkins.

felicepantaleo committed:
remove unused import

cmsbuild commented:
The code-checks are being triggered in jenkins.

kpedro88 commented:
@fwyzard 

> Is there any evidence that it actually gives better performance?
> The proposed change would access the data through a pointer instead of on
the stack.

It would access the object on the heap rather than the stack, but `back()` gives a reference, not a pointer. My intuition is that avoiding the copy is preferable, but indeed, this should be tested empirically.

I also thought we had introduced a clang-tidy check for `const auto&`, but maybe it doesn't catch all cases.

cmsbuild commented:
The code-checks are being triggered in jenkins.

VinInn commented:
> On 23 Oct, 2020, at 2:38 PM, Kevin Pedro <notifications@github.com> wrote:
> 
>  gives a reference, not a pointer
Makes no difference what so ever (just syntactic-sugar)

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31906/19350

- This PR adds an extra 48KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31906/19350/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31906/19350/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31906/19351

- This PR adds an extra 48KB to repository

cmsbuild commented:
Pull request #31906 was updated. @perrotta, @andrius-k, @kmaeshima, @ErnestaP, @kpedro88, @cmsbuild, @jfernan2, @fioriNTU, @slava77, @jpata can you please check and sign again.

felicepantaleo commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.
Test Parameters:
  - **ENABLE_BOT_TESTS** = PROFILING

cmsbuild commented:
+1
Tested at: 933092eca4d26caa4b17846fa76f471d0b35b02e
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-814241/10241/summary.html
CMSSW: CMSSW_11_2_X_2020-10-23-1100
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-814241/10241/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - ROOTFileChecks: Some differences in event products or their sizes found
 - Reco comparison results: 1071 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2544110
 - DQMHistoTests: Total failures: 2389
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2541699
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 34 files compared)
 - Checked 149 log files, 22 edm output root files, 35 DQM output files

kpedro88 commented:
The first profile showed a dramatic decrease in reco CPU usage, driven by ~8% decrease in TrackstersProducer but also small decreases in many other modules:
> Total Cumulative(old->new): 1461.53 -> 1260.69

The second profile shows a less pronounced decrease: still ~7% decrease in TrackstersProducer, but other modules don't change as much
> Total Cumulative(old->new): 1398.08 -> 1316.15

From the changes implemented so far, I can only pick out https://github.com/cms-sw/cmssw/pull/31906/commits/625fa9517bd9c4aa993a10c4aa5b6771b96da89e as one that might impact performance significantly.

However, these profiles were compared to different baseline IBs, and changes in the baseline may have affected the observed performance of this PR.

rovere commented:
@kpedro88 the reduction is solely due to:

* removal of the`MIP` iteration
* selection on the input cluster size

As such, let me stress, the reduction should be considered welcome but, in some sense, **not _permanent_**, since we will review those two topics in the future with the clear goal "to have them back" (in a much better shape wrt the current one, of course).
As a reminder, when in future we will activate them back, we will observe the opposite.

rovere commented:
By the way,
if [this pie-chart](https://cmssdt.cern.ch/circles/web/piechart.php?local=false&dataset=CMSSW_11_2_X_2020-10-23-1100%2Fslc7_amd64_gcc820%2F23434.21%2FPR-814241%2F10241%2Fstep4_PAT_PU.resources&resource=time_thread&colours=.%2Fdefault&groups=.%2Fhlt&threshold=0) has been made using a 200PU workflow, as the name and wfl number suggest, I think this PR alone will also (at least partially) address the increase in size and timing of `miniAOD` that was reported in `pre7` and in the equivalent #31766 backport of #31527 
@SohamBhattacharya @jpata @dpiparo

felicepantaleo commented:
I dropped the energy check before selection as tracksters at that point have energy >0 by construction.

felicepantaleo commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.
Test Parameters:
  - **ENABLE_BOT_TESTS** = PROFILING

cmsbuild commented:
+1
Tested at: 2a5e9ff2359f085db3f88cecccd41f662b260c61
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-814241/10291/summary.html
CMSSW: CMSSW_11_2_X_2020-10-25-2300
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-814241/10291/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - ROOTFileChecks: Some differences in event products or their sizes found
 - Reco comparison results: 1075 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2544110
 - DQMHistoTests: Total failures: 2395
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2541693
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 34 files compared)
 - Checked 149 log files, 22 edm output root files, 35 DQM output files

felicepantaleo committed:
Fix Validation for empty tracksters

cmsbuild commented:
The code-checks are being triggered in jenkins.

felicepantaleo commented:
@cmsbuild please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31906/19412

- This PR adds an extra 192KB to repository

cmsbuild commented:
The tests are being triggered in jenkins.
Test Parameters:
  - **ENABLE_BOT_TESTS** = PROFILING

cmsbuild commented:
Pull request #31906 was updated. @perrotta, @andrius-k, @kmaeshima, @ErnestaP, @kpedro88, @cmsbuild, @jfernan2, @fioriNTU, @slava77, @jpata can you please check and sign again.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-814241/10324/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - ROOTFileChecks: Some differences in event products or their sizes found
 - Reco comparison results: 1016 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2544092
 - DQMHistoTests: Total failures: 2370
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2541700
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 34 files compared)
 - Checked 149 log files, 22 edm output root files, 35 DQM output files

perrotta commented on RecoHGCal/TICL/plugins/TrackstersMergeProducer.cc:
```diff
@@ -168,185 +185,331 @@ void TrackstersMergeProducer::produce(edm::Event &evt, const edm::EventSetup &es
+            for (auto otherTracksterIdx : trackstersTRKwithSameSeed) {
+              tmpCandidate.addTrackster(edm::Ptr<ticl::Trackster>(trackstersMergedHandle, otherTracksterIdx));
+            }
+            float p = tracksterTotalRawPt * cosh(t.barycenter().eta());
```
You are using consistently in sevaral places `p = pt * cosh(vec.eta())`: just for my own education, is it faster than (e.g.) `p = pt * sqrt(1+vec.z()*vec.z()/(vec.x()*vec.x()+(vec.y()*vec.y())` ?

felicepantaleo replied:
of course not, thanks!

perrotta commented on Validation/HGCalValidation/src/HGVHistoProducerAlgo.cc:
```diff
@@ -1511,293 +1511,292 @@ void HGVHistoProducerAlgo::multiClusters_to_CaloParticles(const Histograms& hist
+        for (auto& cpPair : cpsInMultiCluster[mclId]) {
+          //In case of a multi cluster with zero energy but related CaloParticles the score is set to 1.
+          cpPair.second = 1.;
+          // LogDebug("HGCalValidator") << "multiCluster Id: \t" << mclId
```
This is repeated below, and the commented out part can be removed

felicepantaleo committed:
modify comment to reflect code

felicepantaleo committed:
replace parameter root_doublet_max_distance_from_seed with its square

felicepantaleo committed:
Optimization

felicepantaleo committed:
remove unused parameter

felicepantaleo committed:
remove commented out code

felicepantaleo committed:
optimization

felicepantaleo committed:
code format

felicepantaleo committed:
code format

felicepantaleo committed:
code format

cmsbuild commented:
The code-checks are being triggered in jenkins.

felicepantaleo commented on RecoHGCal/TICL/plugins/PatternRecognitionbyCA.cc:
```diff
@@ -137,29 +153,102 @@ void PatternRecognitionbyCA<TILES>::makeTracksters(
+      for (const auto &layer : uniqueLayerIds) {
+        if (layer != j) {
+          numberOfMissingLayers++;
+          j++;
```
yes @perrotta this is the desired behavior

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31906/19429

- This PR adds an extra 204KB to repository

cmsbuild commented:
Pull request #31906 was updated. @perrotta, @andrius-k, @kmaeshima, @ErnestaP, @kpedro88, @cmsbuild, @jfernan2, @fioriNTU, @slava77, @jpata can you please check and sign again.

felicepantaleo commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.
Test Parameters:
  - **ENABLE_BOT_TESTS** = PROFILING

kpedro88 commented:
@smuzaffar the tests appear to have timed out. is it possible to extend the timeout automatically when profiling is enabled?

smuzaffar commented:
@kpedro88 job timed out after 24 hours. I think running PR tests for more than 24 hours is not suitable. Many tokens gets expired and we will start seeing other issues :-)

@gartung, can you please check why profiling is taking over 20 hours to run?

gartung commented:
The timeout on the profiling job is 4 hours not 24 hours.

felicepantaleo committed:
optimization

kpedro88 commented:
please test
i would be surprised if this is a "real" effect (i.e. caused by the PR, rather than some issue with the build machine or another transient condition)

cmsbuild commented:
The tests are being triggered in jenkins.
Test Parameters:
  - **ENABLE_BOT_TESTS** = PROFILING

felicepantaleo committed:
Simplify energy assignment

felicepantaleo committed:
Provide EcalEnergy and HcalEnergy to JetMET for calibration

cmsbuild commented:
The code-checks are being triggered in jenkins.

felicepantaleo committed:
code format

cmsbuild commented:
The code-checks are being triggered in jenkins.

felicepantaleo commented:
@cmsbuild please test

perrotta commented on RecoHGCal/TICL/plugins/TrackstersMergeProducer.cc:
```diff
@@ -390,45 +389,49 @@ void TrackstersMergeProducer::produce(edm::Event &evt, const edm::EventSetup &es
-            usedTrackstersMerged[otherTracksterIdx] = true;
           }
           tracksterTotalRawPt += trackstersMergedHandle->at(closestTrackster).raw_pt() - t.raw_pt();
+          usedTrackstersMerged[closestTrackster] = true;
```
Is this a bug fix? Should we expect any effect on the results (i.e., should you update the validation plots?)

felicepantaleo replied:
yes, the results in PF validation should change now, we are rerunning it

cmsbuild commented:
+1
Tested at: 55110b5d2590b3f1b0022e65a32eef4fbdcb8362
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-814241/10409/summary.html
CMSSW: CMSSW_11_2_X_2020-10-30-1100
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-814241/10409/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - ROOTFileChecks: Some differences in event products or their sizes found
 - Reco comparison results: 1012 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2544092
 - DQMHistoTests: Total failures: 2364
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2541706
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 34 files compared)
 - Checked 149 log files, 22 edm output root files, 35 DQM output files

gartung commented:
@smuzaffar The profiling completed the second time it was run/

felicepantaleo committed:
assign track momentum direction at point of closest approach instead of trackster's barycenter in case of charged particles

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31906/19497

- This PR adds an extra 228KB to repository

- Found files with invalid states:
  - DQM/BeamMonitor/plugins/OnlineBeamMonitor.cc.orig:
    - Added: 50576c0014a67777ae2ce7ea536a4bc3d92d0d13
    - Deleted: ffcf25a5bd9eef4e2d68b6efaeb198bbf1237694
  - DQM/BeamMonitor/plugins/OnlineBeamMonitor.h.orig:
    - Added: 50576c0014a67777ae2ce7ea536a4bc3d92d0d13
    - Deleted: ffcf25a5bd9eef4e2d68b6efaeb198bbf1237694
  - DQM/BeamMonitor/plugins/OnlineBeamMonitor.cc.rej:
    - Added: 50576c0014a67777ae2ce7ea536a4bc3d92d0d13
    - Deleted: ffcf25a5bd9eef4e2d68b6efaeb198bbf1237694


- There are other open Pull requests which might conflict with changes you have proposed:
  - File  CommonTools/PileupAlgos/python/Puppi_cff.py  modified in PR(s): #31893
  - File  Configuration/Geometry/python/dict2026Geometry.py  modified in PR(s): #30126
  - File  Configuration/PyReleaseValidation/python/relval_steps.py  modified in PR(s): #31795, #31854
  - File  Configuration/PyReleaseValidation/python/upgradeWorkflowComponents.py  modified in PR(s): #31726, #31795, #31854
  - File  Configuration/StandardSequences/python/Eras.py  modified in PR(s): #31726
  - File  DQM/Integration/python/clients/pixel_dqm_sourceclient-live_cfg.py  modified in PR(s): #27983, #31721
  - File  DQM/Integration/python/clients/sistrip_dqm_sourceclient-live_cfg.py  modified in PR(s): #27983, #31721
  - File  Geometry/HGCalCommonData/interface/HGCalTypes.h  modified in PR(s): #31994
  - File  Geometry/HGCalCommonData/src/HGCalGeomParameters.cc  modified in PR(s): #31994
  - File  Geometry/HGCalCommonData/src/HGCalWaferMask.cc  modified in PR(s): #31994
  - File  Geometry/HGCalGeometry/test/HGCalWaferInFileTest.cc  modified in PR(s): #31994
  - File  L1Trigger/CSCTriggerPrimitives/BuildFile.xml  modified in PR(s): #31631
  - File  PhysicsTools/NanoAOD/plugins/GenWeightsTableProducer.cc  modified in PR(s): #31982, #31680, #31946
  - File  PhysicsTools/NanoAOD/python/electrons_cff.py  modified in PR(s): #31974
  - File  PhysicsTools/NanoAOD/python/nanoDQM_cfi.py  modified in PR(s): #31531
  - File  PhysicsTools/PatAlgos/python/slimming/MicroEventContent_cff.py  modified in PR(s): #31220, #31985, #31935, #31971
  - File  PhysicsTools/PatAlgos/python/slimming/slimming_cff.py  modified in PR(s): #31220, #31935
  - File  RecoTauTag/RecoTau/python/tools/runTauIdMVA.py  modified in PR(s): #31744
  - File  RecoTracker/TkSeedGenerator/plugins/BuildFile.xml  modified in PR(s): #27983, #31722
  - File  RecoVertex/PrimaryVertexProducer/python/OfflinePrimaryVertices_cfi.py  modified in PR(s): #31935
  - File  SimTracker/SiPhase2Digitizer/plugins/Pixel3DDigitizerAlgorithm.cc  modified in PR(s): #31697
  - File  SimTracker/SiPhase2Digitizer/plugins/Pixel3DDigitizerAlgorithm.h  modified in PR(s): #31697


Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-checks**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31906/19497/code-checks.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31906/19497/code-checks.patch | patch -p1`
You can also run `scram build code-checks` to apply code checks directly

cmsbuild commented:
The code-checks are being triggered in jenkins.

felicepantaleo commented:
rebased on a more recent IB and modified the direction of the Candidate based on the track information (when available) instead of the barycenter. Thanks for the feedback for the PF Validation @hqucms

felicepantaleo commented:
@cmsbuild please test

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31906/19498

- This PR adds an extra 76KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-checks**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31906/19498/code-checks.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31906/19498/code-checks.patch | patch -p1`
You can also run `scram build code-checks` to apply code checks directly

felicepantaleo committed:
avoid copy

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31906/19499

- This PR adds an extra 76KB to repository

cmsbuild commented:
Pull request #31906 was updated. @perrotta, @andrius-k, @kmaeshima, @ErnestaP, @kpedro88, @cmsbuild, @jfernan2, @fioriNTU, @slava77, @jpata can you please check and sign again.

felicepantaleo commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.
Test Parameters:
  - **ENABLE_BOT_TESTS** = PROFILING

cmsbuild commented:
+1
Tested at: abc84da0735ccd850b313111396a8ec7361f4423
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-814241/10425/summary.html
CMSSW: CMSSW_11_2_X_2020-10-31-1100
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-814241/10425/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - ROOTFileChecks: Some differences in event products or their sizes found
 - Reco comparison results: 1016 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2544092
 - DQMHistoTests: Total failures: 2370
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2541700
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 34 files compared)
 - Checked 149 log files, 22 edm output root files, 35 DQM output files

jfernan2 commented:
@felicepantaleo there are as many as 1k MEs modified in this PR in the 2026 WFs, even in the EGamma folder.
I understand these are expected due to the introduction of the track seeded EM iteration.
https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_11_2_X_2020-10-31-1100+814241/39674/dqm-histo-comparison-summary.html
Thanks

jfernan2 commented:
In addtion, as commented in the backport, with this PR plots in HGCAL / TICLPFCandidates / 3 get now zero entries and HGCAL / HGCalValidator / ticlMultiClustersFromTrackstersMerge/Trk now get many Denominator null values
Is this expected understood?
Thanks

jfernan2 commented:
Please ignore my last comment, it has been clarified on the backport comment:
https://github.com/cms-sw/cmssw/pull/31907#issuecomment-720476044

jfernan2 commented:
+1

felicepantaleo commented:
> @felicepantaleo will you be able to explore if using `emplace_back()` and `back()` in `TrackstersMergeProducer` (and one in `PatternRecognitionbyCA`) is preferable for CPU usage compared to copying temporary objects? (The CPU usage here is already quite reasonable, but I'm still curious.)

@kpedro88 sorry for the late reply, I just did and copying temporary object from the stack is slightly faster

kpedro88 commented:
Good to know, thanks.

felicepantaleo committed:
Add missing const.

felicepantaleo committed:
Make track cuts configurable

cmsbuild commented:
The code-checks are being triggered in jenkins.

felicepantaleo committed:
code format

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31906/19524

- This PR adds an extra 20KB to repository

cmsbuild commented:
Pull request #31906 was updated. @perrotta, @andrius-k, @kmaeshima, @ErnestaP, @kpedro88, @cmsbuild, @jfernan2, @fioriNTU, @slava77, @jpata can you please check and sign again.

felicepantaleo commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.
Test Parameters:
  - **ENABLE_BOT_TESTS** = PROFILING

felicepantaleo commented on RecoHGCal/TICL/plugins/PatternRecognitionbyCA.cc:
```diff
@@ -137,29 +155,105 @@ void PatternRecognitionbyCA<TILES>::makeTracksters(
+      //if a seeding region does not lead to any trackster
+      tmp.setSeed(input.regions[0].collectionID, seedIndices[tracksterId]);
+
+      std::pair<float, float> timeTrackster(-99., -1.);
```
@kpedro88 I would leave the code as it is for the moment as it was not changed by the PR and timing is being restructured in a PR that will come right after this is merged.

felicepantaleo replied:
@amartelli please take note of this comment

cmsbuild commented:
+1
Tested at: bbbacf1c98af9f3e974b1291f1ef521921c8fb7d
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-814241/10449/summary.html
CMSSW: CMSSW_11_2_X_2020-11-02-1100
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-814241/10449/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - ROOTFileChecks: Some differences in event products or their sizes found
 - Reco comparison results: 1016 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2544092
 - DQMHistoTests: Total failures: 2370
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2541700
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 34 files compared)
 - Checked 149 log files, 22 edm output root files, 35 DQM output files

mariadalfonso commented on RecoHGCal/TICL/plugins/TrackstersMergeProducer.cc:
```diff
@@ -499,11 +507,12 @@ void TrackstersMergeProducer::produce(edm::Event &evt, const edm::EventSetup &es
     auto const &track = tracks[i];
-    if (track.pt() > 1.f and track.quality(reco::TrackBase::highPurity) and track.missingOuterHits() < 5 and
-        std::abs(track.outerEta()) > 1.48 and std::abs(track.outerEta()) < 3.0 and usedSeeds[i] == false) {
+    if (track.pt() > track_min_pt_ and track.quality(reco::TrackBase::highPurity) and track_max_missing_outerhits_ < 5 and
```
track_max_missing_outerhits_ < 5 
should be 
track.missingOuterHits() < track_max_missing_outerhits_

felicepantaleo replied:
thanks @mariadalfonso , you're right, fixed

felicepantaleo committed:
fix configurable parameter

cmsbuild commented:
The code-checks are being triggered in jenkins.

felicepantaleo commented:
@cmsbuild please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31906/19542

- This PR adds an extra 20KB to repository

cmsbuild commented:
The tests are being triggered in jenkins.
Test Parameters:
  - **ENABLE_BOT_TESTS** = PROFILING

cmsbuild commented:
Pull request #31906 was updated. @perrotta, @andrius-k, @kmaeshima, @ErnestaP, @kpedro88, @cmsbuild, @jfernan2, @fioriNTU, @slava77, @jpata can you please check and sign again.

cmsbuild commented:
+1
Tested at: 92c59aac9e13ac34526f972536a03747009b0680
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-814241/10466/summary.html
CMSSW: CMSSW_11_2_X_2020-11-02-2300
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-814241/10466/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - ROOTFileChecks: Some differences in event products or their sizes found
 - Reco comparison results: 1012 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2544092
 - DQMHistoTests: Total failures: 2364
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2541706
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 34 files compared)
 - Checked 149 log files, 22 edm output root files, 35 DQM output files

jfernan2 commented:
+1

felicepantaleo commented:
you can find the latest physics results with this pull request described in the talks by HGCAL and Jets/MET at the HLT Upgrade workshop:
https://indico.cern.ch/event/962025/#4-hgcal
https://indico.cern.ch/event/962025/#6-jetsmet

perrotta commented:
> you can find the latest physics results with this pull request described in the talks by HGCAL and Jets/MET at the HLT Upgrade workshop:
> https://indico.cern.ch/event/962025/#4-hgcaluld 
> https://indico.cern.ch/event/962025/#6-jetsmet

Thank you @felicepantaleo 
The plots shown at the TSG workshop correspond to commit abc84da, which is fine because changes interveaned afterwards are not relevant for the final validation results.
By the way, would you be able to update also the figures in the PR description with the output after the latest (relevant) commits, for documentation purposes?

felicepantaleo commented:
done

missirol commented:
> you can find the latest physics results with this pull request described in the talks by HGCAL and Jets/MET at the HLT Upgrade workshop

> The plots shown at the TSG workshop correspond to commit abc84da


One minor correction: the HLT-JME plots linked above by Felice include updates up to 4da60d876530939e2e9867ac5da572642430f2cc (from the `11_1_X` backport in #31907); I think that would correspond to 6c32c60c232982470d20b398ee1b203bd5d6f37a in this PR.

perrotta commented:
> da60d8

Thank you @missirol.

There were several updates and bug fixes applied to this PR, and I think some final check and validation should be allowed based on the very final version before merging it.  Also the plots in the PR description seems to correspond to a supposed 9b855f6 commit, which I am not able to retrieve neither here or in the backport PR.

I expect that the authors and the HGCal team can provide some kind of green light based on it. The same for the performance, which should be better re-computed with the very final version (by the way, has this PR settled down by now?)

rovere commented:
> > you can find the latest physics results with this pull request described in the talks by HGCAL and Jets/MET at the HLT Upgrade workshop
> 
> > The plots shown at the TSG workshop correspond to commit [abc84da](https://github.com/cms-sw/cmssw/commit/abc84da0735ccd850b313111396a8ec7361f4423)
> 
> One minor correction: the HLT-JME plots linked above by Felice include updates up to [4da60d8](https://github.com/cms-sw/cmssw/commit/4da60d876530939e2e9867ac5da572642430f2cc) (from the `11_1_X` backport in #31907); I think that would correspond to [6c32c60](https://github.com/cms-sw/cmssw/commit/6c32c60c232982470d20b398ee1b203bd5d6f37a) in this PR.

Thanks @missirol 
The commits that were added after the one you pointed out in #31907 are related to the various technical changes requested while reviewing this PR, the addition of the `EM` and `HAD` information to the `pfCandidates` produced by `TICL` and, lastly, the correction on the direction taking the track information as input, when available.
I guess you are in the process of testing the `EM` and `HAD` assignment in order to derive `JEC` in`11_1`.
In any case, I do not expect dramatic changes to the overall reconstruction even after the inclusion of the latest commits.
There are still a couple of commits missing in the backport, but again those are mainly technical, with no impact on the physics.
I'll wait up until this PR  gets merged to have them backported too in `11_1`.

perrotta commented:
Testing with just 20 events from the wf 23234.0 (TTbar with 2026 D49 geometry) the overall event size reductions are
- 1.3% for the FEVTDEBUGHLT event content: ```6481072 ->     6396632     -84440            -1.3     ALL BRANCHES ```
- 6.4% for the MINIAODSIM event content:      ```70405 ->       65869      -4536            -6.4     ALL BRANCHES```

perrotta commented:
@slava77 the new products created in output by this PR are the following:
```
      0.0 ->       809.1        809     NEWO   0.01     TICLCandidates_ticlTrackstersMerge__RECO.
      0.0 ->       616.5        616     NEWO   0.01     ticlTracksters_ticlTrackstersTrkEM__RECO.
      0.0 ->      1123.3       1123     NEWO   0.02     recoHGCalMultiClusters_ticlMultiClustersFromTrackstersTrkEM__RECO.
      0.0 ->       344.2        344     NEWO   0.01     floats_ticlTrackstersTrkEM__RECO.
```

perrotta commented:
Trying to summarize the discussion happened yestreday in this thread: please let me know whether you intend to provide updated validations and comparisons, and if so when they can be ready, so that the review can get finalized here.

felicepantaleo commented:
@perrotta no further validation would be produced in this PR. I think you can proceed with the finalization of the review.

kpedro88 commented:
+upgrade

perrotta commented on RecoHGCal/TICL/plugins/PatternRecognitionbyCA.cc:
```diff
@@ -26,9 +26,19 @@ PatternRecognitionbyCA<TILES>::PatternRecognitionbyCA(const edm::ParameterSet &c
-      min_clusters_per_ntuplet_(conf.getParameter<int>("min_clusters_per_ntuplet")),
+      skip_layers_(conf.getParameter<int>("skip_layers")),
+      max_missing_layers_in_trackster_(conf.getParameter<int>("max_missing_layers_in_trackster")),
+      check_missing_layers_(max_missing_layers_in_trackster_ < 100),
```
I would have found more intuitutive using a negative number to identify the max number which corresponds to "do not check". But ok, the behaviour doesn't change anyhow

perrotta commented:
+1
- TICL algo upgraded as described, with clear improvements in the electron energy resolution and updates in the PF description
- Event size shrinks (approx 6% less in miniAOD)
- Timing also gets reduced, but this is said to be a temporary effect due to the removal of theMIP iteration and selection on the input cluster size: both them are planned to be reintroduced as soon as dedicated studies will be carried on (and the cpu performance will get affected consequently)
- Jenkins tests pass and the differences observed there are consequence of the updates implemented, including the removal of a few objects from the event content (including the `CandidateFromTracksters` and `TrackstersMIP`)

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @silviodonato, @dpiparo, @qliphy (and backports should be raised in the release meeting by the corresponding L2)

silviodonato commented:
As expected we see changes in the Phase-2 workflows
```
23234.0_TTbar_14TeV+2026D49+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14+DigiTrigger+RecoGlobal+HARVESTGlobal
23434.999_TTbar_14TeV+2026D49PU_PMXS1S2PR+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14+PREMIX_PremixHLBeamSpot14PU+DigiTriggerPU+RecoGlobalPU+HARVESTGlobalPU
28234.0_TTbar_14TeV+2026D60+TTbar_14TeV_TuneCP5_GenSimHLBeamSpot14+DigiTrigger+RecoGlobal+HARVESTGlobal
```
Do you know why we see many more difference in `23234.0` and `28234.0`? Since this PR improves the performance against the pile-up, I expected to see larger differences in `23434.999` which has pileup instead of `23234.0` and `28234.0` which are without pileup.


Another question: are these (small) differences in the tau ID expected?
https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_11_2_X_2020-11-02-2300+814241/39695/validateJR/all_mini_OldVSNew_TTbar14TeV2026D49wf23234p0/
https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_11_2_X_2020-11-02-2300+814241/39695/validateJR/all_mini_OldVSNew_TTbar14TeV2026D60wf28234p0/

slava77 commented:
> I expected to see larger differences in `23434.999`

I'm beginning to suspect that this workflow is broken in some way
@kpedro88 did we have any physics validation for it?

kpedro88 commented:
I'm also suspicious that something funny is happening with `23434.999`. I don't see anything out of the ordinary in the matrix test logs. I probably won't have time to look into this in more detail in the next few days; maybe it's worth opening an issue to keep track.

silviodonato commented:
+1

perrotta commented on RecoHGCal/TICL/plugins/TrackstersMergeProducer.cc:
```diff
@@ -168,185 +190,345 @@ void TrackstersMergeProducer::produce(edm::Event &evt, const edm::EventSetup &es
+            auto thisPt = tracksterTotalRawPt + trackstersMergedHandle->at(otherTracksterIdx).raw_pt() - t.raw_pt();
+            closestTrackster = std::abs(thisPt - track.pt()) < minPtDiff ? otherTracksterIdx : closestTrackster;
+          }
+          tracksterTotalRawPt += trackstersMergedHandle->at(closestTrackster).raw_pt() - t.raw_pt();
```
@felicepantaleo this line (also pointed out by the static analyzer) escaped my review of the PR: this increment is completely useless here. Either the line can/should be removed, or it was originally intended to do something different and it has to be fixed then. Please check and provide the fix.

felicepantaleo replied:
thanks @perrotta that line can be erased safely. I will make a pr now

slava77 commented:
> @slava77 the new products created in output by this PR are the following:
> 
> ```
>       0.0 ->       809.1        809     NEWO   0.01     TICLCandidates_ticlTrackstersMerge__RECO.
>       0.0 ->       616.5        616     NEWO   0.01     ticlTracksters_ticlTrackstersTrkEM__RECO.
>       0.0 ->      1123.3       1123     NEWO   0.02     recoHGCalMultiClusters_ticlMultiClustersFromTrackstersTrkEM__RECO.
>       0.0 ->       344.2        344     NEWO   0.01     floats_ticlTrackstersTrkEM__RECO.
> ```

reco monitoring now covers these (sorry, it took a while for me to get to updating the script).
